pkg/log/logger.go

```diff
@@ -23,50 +23,63 @@ import (
 	"fmt"
 	"reflect"
 	"sync"
+	"sync/atomic"
 
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
 )
 
 var once sync.Once
-var logger *zap.Logger
-var config *zap.Config
+var holderValue atomic.Value
+
+type loggerHolder struct {
+	logger     *zap.Logger
+	zapConfigs *zap.Config
+}
 
 func Logger() *zap.Logger {
-	once.Do(func() {
-		if logger = zap.L(); isNopLogger(logger) {
-			// If a global logger is not found, this could be either scheduler-core
-			// is running as a deployment mode, or running with another non-go code
-			// shim. In this case, we need to create our own logger.
-			// TODO support log options when a global logger is not there
-			config = createConfig()
-			var err error
-			logger, err = config.Build()
-			// this should really not happen so just write to stdout and set a Nop logger
-			if err != nil {
-				fmt.Printf("Logging disabled, logger init failed with error: %v\n", err)
-				logger = zap.NewNop()
-			}
+	once.Do(initLogger)
+	return holderValue.Load().(loggerHolder).logger
+}
+
+func GetConfig() *zap.Config {
+	once.Do(initLogger)
+	return holderValue.Load().(loggerHolder).zapConfigs
+}
+
+func initLogger() {
+	var logger *zap.Logger
+	var config *zap.Config
+
+	if logger = zap.L(); isNopLogger(logger) {
+		// If a global logger is not found, this could be either scheduler-core
+		// is running as a deployment mode, or running with another non-go code
+		// shim. In this case, we need to create our own logger.
+		config = createConfig()
+		var err error
+		logger, err = config.Build()
+		// this should really not happen so just write to stdout and set a Nop logger
+		if err != nil {
+			fmt.Printf("Logging disabled, logger init failed with error: %v\n", err)
+			logger = zap.NewNop()
 		}
+	}
+	holderValue.Store(loggerHolder{
+		logger:     logger,
+		zapConfigs: config,
 	})
-
-	return logger
 }
 
 func InitializeLogger(log *zap.Logger, zapConfig *zap.Config) {
-	once.Do(func() {
-		logger = log
-		config = zapConfig
-		logger.Info("Using an already initialized logger")
+	once.Do(initLogger)
```

Comment:
This changes the init for the logger from the shim into the core. The `InitializeLogger()` passes in a pre-build config from the k8shim and makes sure that the shim and the core run with *the exact same* config object. Without that changing the log level via the rest call is only going to change the core log level.

It also does the work twice in this case: we have a logger and config we just want to set it. We can change that in a later jira to decouple things but not as part of this change.

Reply:
initLogger is used now only to provide a "fallback" logger in case something is logged early. Once the scheduler is initialized propery, the dynamic configuration will replace the logger with the real, configured one. This is all part of allowing dynamic reconfiguration via the configmap.

Reply:
Per offline discussions, reverting all logging changes.